x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/SparseFileTracker.java

```diff
@@ -464,8 +474,20 @@ private void onGapSuccess(final Range gapRange) {
     private void maybeUpdateCompletePointer(Range gapRange) {
         assert Thread.holdsLock(ranges);
         if (gapRange.start == 0) {
-            assert complete <= gapRange.end;
-            complete = gapRange.end;
+            updateCompletePointerHoldingLock(gapRange.end);
+        }
+    }
+
+    private void updateCompletePointerHoldingLock(long value) {
+        assert Thread.holdsLock(ranges);
+        assert complete <= value : complete + ">" + value;
```

Comment:
When called from `maybeUpdateCompletePointer`, this assertion depends on `ProgressListenableActionFuture` not notifying of progress on the last byte. It feels slightly subtle to me, but works ofc. The check I suggested in `maybeUpdateCompletePointer` should fix it though.